### PR TITLE
Add Apple Metal codegen backend.

### PR DESCRIPTION
**Pull request by zvookin**
This PR provides code generation for Apple's Metal GPU compute API. Support is generally at the same level as for OpenCL or PTX, which means almost of of Halide works. Metal is supported on iOS and the next version of Mac OS X.

A test app based on the HelloIOS app is included. This test app endeavors to do all its computation on the GPU and then to blit directly to the texture of a Metal layer. Performance is rather dreadful. On a first generation iPad Air, the CPU version of the reaction diffusion app achieves a frame time of 17ms while the Metal version takes 430ms. I have done some basic investigation and have not found an obvious issue in the code generated by Halide, but this is clearly something to be worked on. I'd like to get the PR moved forward so others can look at and work with the code and possibly make progress on performance.

### CODE REVIEW

File: src/CodeGen_Metal_Dev.cpp [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+namespace {
+
+// If e is a ramp expression with stride 1, return the base, otherwise undefined.
+Expr is_ramp1(Expr e) {
+    const Ramp *r = e.as<Ramp>();
+    if (r == NULL) {
+        return Expr();
+    }
+
+    const IntImm *i = r->stride.as<IntImm>();
```
Comment (abadams) [inline src/CodeGen_Metal_Dev.cpp]: if (is_one(r->stride))

File: src/runtime/HalideRuntimeMetal.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ *  valid on an Metal device, or not have any associated device
+ *  memory. If there is no device memory (dev field is NULL), this
+ *  returns 0.
+ */
+extern uintptr_t halide_metal_get_buffer(void *user_context, struct buffer_t *buf);
+
+struct halide_metal_device;
+struct halide_metal_command_queue;
+
+// TODO: docs, check if these are exported for other GPU APIs.
```
Comment (abadams) [inline src/runtime/HalideRuntimeMetal.h]: There are a couple of doc TODOs still here.

**Comment by abadams** [comment]
Do you think it's possible to do automated tests of this? I have a recent ipad that I could plug into the mac build bot. I'm not sure how people do continuous integration for ios apps.

**Comment by zvookin** [comment]
The best way to do automated tests is on a Mac running the next version of
Mac OS X (El Capitan). Metal is supported on the desktop in that OS. (This
is public information, but that OS is still in beta.) All of the JIT tests
worked last I tested and I will run them again before merge.

It would be great to get automated testing at all on Android and iOS, but
doing so requires solving the JIT vs. AOT issue in tests. (For Android we
could run JIT. It is basically impossible on iOS I believe.)

**Comment by abadams** [comment]
I wonder if we can at least determine if the sample ios app starts and runs
without crashing.

**Comment by zvookin** [comment]
This should be ready to merge. Passes all tests on Mac OS X. (After latest set of changes of course.) The CMake support does not set up the linker flags to include the Metal and Foundation frameworks so that will not work. I glanced at it and decided someone else will have to fix that if it matters.

I am not convinced the performance issues on iOS are due to codegen or GPU speed so much as weird throttling either due to the application structure waiting inordinately or Metal not allowing compute shaders to use lots of bandwidth on the GPU in order to preserve responsiveness on the device. In any event I think the quality of implementation of the Metal backend is pretty much on par with our OpenCL and CUDA ones.

**Comment by delcypher** [comment]
> The CMake support does not set up the linker flags to include the Metal and Foundation frameworks so that will not work. I glanced at it and decided someone else will have to fix that if it matters.

The CMake build is my daily driver so this needs to be fixed, also TravisCI will not be happy until it is. I'll take a look and see if I can reproduce your makefile changes.

At a glance I noticed you have a git submodule for LLVM in your branch. I don't think that should be there unless you get agreement from all the Halide devs that we want this.

**Comment by delcypher** [comment]
Note: I don't have OSX with metal support so I can't test this but at the very least I can unbreak the CMake build under linux.

**Comment by zvookin** [comment]
CMake Metal support is only needed if one is trying to generate and run
code for Metal, which only happens in build context if one tries to run the
tests with HL_JIT_TARGET=host-metal or similar. I.e. it should not break
anything to not have the frameworks needed for Metal in the CMake build. We
cannot test Metal on any test setup we have at present as it requires a
beta version of Mac OS X.

I thought I deleted the submodule change. That was there because git kept
complaining about being untracked or some such and is largely an accident.
I will double check to make sure it is not in the branch.

**Comment by delcypher** [comment]
@zvookin
The CMake build was broken [see here](https://travis-ci.org/halide/Halide/builds/76787303)

I pushed some changes that should fix the CMake build under linux and also changed how the metal runtime is handled in ``LLVM_Runtime_Linker.cpp`` so that ``WITH_METAL`` is actually used.

A few other comments

* Remnants of the LLVM submodule are still there. Temporary stuff like this probably should not be merged into the master branch. You could probably do a rebase to fix this and also clean up your commits (the e-mail and hostname of some of them are odd which will be unhelpful if someone tries to do a ``git blame``) and drop the "merge with master" commits.

* It occurs to me that the disabling of a target with the ``TARGET_*`` option in CMake (and the ``WITH_*``) is a little bit odd because if a target is disabled we still compile the runtime for it (e.g. ``TARGET_OPENGL=OFF``) anyway.

File: llvm [hunk 3]
```diff
@@ -0,0 +1 @@
+Subproject commit 327cfdf4fe77a2fc0e005b20bdc5980d476ef218
```
Comment (abadams) [inline llvm]: I think this is the submodule that needs removing.

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2.cpp [hunk 4]
```diff
@@ -105,6 +105,7 @@ int main(int argc, char **argv) {
 
         new_state.vectorize(x, 4);
         blur.vectorize(x, 4);
+	state.set_bounds(2, 0, 3);
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2.cpp]: tab character here

**Comment by delcypher** [comment]
> Wow, this was pretty broken. I guess we didn't run with any initmods disabled in any of the tests!

@abadams Related to this, it seems the runtime for some targets still get built and bundled into libHalide even if they are disabled via ``TARGET_*`` (e.g. OpenGL). Is this intentional?

**Comment by abadams** [comment]
No, that's not intentional. It doesn't matter much though.

**Comment by zvookin** [comment]
I believe the llvm submodule should be removed from this merge now. Tab in reaction_diffusion.cpp also fixed and merged up to master.

**Comment by zvookin** [comment]
Re: how WITH_* flags work, these are mostly there to make Halide smaller if one is not using certain backends, e.g. in a self-contained system doing JIT for one platform, and ideally to make it possible to remove the dependency on LLVM entirely for certain use cases. There are no uses I know of at present that exercise either of these use cases so I'm not surprised this was broken. We might add a test to make sure the system builds with various combinations of the WITH_* flags I guess.

**Comment by abadams** [comment]
The travis failures seem to be timeouts connecting to llvm.org, rather than an actual bug. Given that the changes since the last travis failure are trivial, I'll just merge this.

It's nice to keep git history clean, but I don't think sanitizing it after-the-fact is a good use of time.

**Pull request merged**